jormungandr/src/network/p2p/topology.rs

```diff
@@ -126,8 +139,11 @@ impl P2pTopology {
         from: Id,
         gossips: Gossips,
     ) -> impl Future<Item = (), Error = E> {
-        self.write()
-            .map(move |mut topology| topology.accept_gossips(from.into(), gossips.into()))
+        self.write().map(move |mut topology| {
+            if topology.nodes().node_count().all_count < DEFAULT_MAX_NODES {
+                topology.accept_gossips(from.into(), gossips.into())
+            }
```

Comment:
I would rather always admit the newly advertised entries after freeing the requisite number of older ones. It does not matter how the nodes are picked for removal now; might just use the collection iterator.

Reply:
This may need changes in `poldercast`. I suggest implementing a configurable size limit and eviction functionality inside `poldercast::Topology`, so that we can fine-tune it later, possibly using the global node timestamps.
Cc @NicolasDP

Reply:
> I would rather always admit the newly advertised entries after freeing the requisite number of older ones. It does not matter how the nodes are picked for removal now; might just use the collection iterator.

It's kind of arbitrary when you think about it. It's just easier to let the natural eviction process work to purge nodes as they are attempted to be gossiped with. That would of course mean a change to the PolderCast library rather than Jormungandr.

> This may need changes in `poldercast`. I suggest implementing a configurable size limit and eviction functionality inside `poldercast::Topology`, so that we can fine-tune it later, possibly using the global node timestamps.

Yep. I think you are right on the money here. That's precisely why I decided to make the change Jormungandr-side. I initially started to implement this in PolderCast create but quickly realised a few things:

1. There really needs to be that valance/timestamp value propagated on the network that you [spoke about here](https://github.com/primetype/poldercast/issues/10). No way that change would make it in for the next release in time.
2. More important that this pool nodes are stabilised. This problem is feeding itself; causing nodes to restart all the time with new IDs, only for those to be propagated forever. We can slow that down by relieving processing on the nodes.
3. I acknowledge this is not a long term solution. But its minimal code change for maximum effect. Low risk, high reward (I hope).

Reply:
While testing this PR, I've also observed a significant reduction in CPU consumption while maintaining a flat utilization curve.  Others have reported the same behavior as well as building and propagating their blocks.  I have observed a slightly lower "Established" TCP connection count in my configuration.

Reply:
As nodes in peerAvailableCnt restart or disconnect they are Quarantined and over time peerQuarantinedCnt might grow to be a larger count than peerAvailableCnt.

I'm not sure if GC is flushing peerQuarantinedCnt fast enough if at all to allow fresh connections when setting a hard limit. Ultimately you may need to restart, though definitely a step in the right direction

Reply:
> I would rather always admit the newly advertised entries after freeing the requisite number of older ones. It does not matter how the nodes are picked for removal now; might just use the collection iterator.

I agree with this. I have made some changes in my implementation of _poldercast_. `0.12` adds the limits of the overall peer collections based using an LRU. This allows for our local node to ignore more easily nodes that are not of used or that have been quarantined under long network churn time (which may happen when new versions are being propagated to the network -- every nodes rebooting to newer versions and propagating new gossips).

Reply:
Sounds good. I'll close out this issue in light of the upstream changes in PolderCast.